packages/xrpl/src/models/transactions/oracleSet.ts

```diff
@@ -0,0 +1,91 @@
+import { ValidationError } from '../../errors'
+
+import {
+  BaseTransaction,
+  isNumber,
+  isString,
+  validateBaseTransaction,
+  validateOptionalField,
+  validateRequiredField,
+} from './common'
+
+// TODO: add docs
+
+export interface PriceData {
+  BaseAsset: string
+  QuoteAsset: string
+  AssetPrice?: number
+  Scale?: number
+}
+
+/**
+ * @category Transaction Models
+ */
+export interface OracleSet extends BaseTransaction {
+  TransactionType: 'OracleSet'
+  OracleDocumentID: number
+  LastUpdateTime: number
+  Provider?: string
+  URI?: string
+  AssetClass?: string
+  PriceDataSeries?: PriceData[]
+}
+
+/**
+ * Verify the form and type of a OracleSet at runtime.
+ *
+ * @param tx - A OracleSet Transaction.
+ * @throws When the OracleSet is malformed.
+ */
+export function validateOracleSet(tx: Record<string, unknown>): void {
+  validateBaseTransaction(tx)
+
+  validateRequiredField(tx, 'OracleDocumentID', isNumber)
+
+  validateRequiredField(tx, 'LastUpdateTime', isNumber)
+
+  validateOptionalField(tx, 'Provider', isString)
+
+  validateOptionalField(tx, 'URI', isString)
+
+  validateOptionalField(tx, 'AssetClass', isString)
+
+  validateOptionalField(tx, 'PriceDataSeries', (value) => {
+    if (!Array.isArray(value)) {
+      throw new ValidationError('OracleSet: PriceDataSeries must be an array')
+    }
+
+    for (const priceData of value) {
+      if (typeof priceData !== 'object') {
+        throw new ValidationError(
+          'OracleSet: PriceDataSeries must be an array of objects',
+        )
+      }
+
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access -- we are validating the type
+      if (typeof priceData.BaseAsset !== 'string') {
+        throw new ValidationError(
+          'OracleSet: PriceDataSeries must have a `BaseAsset` string',
+        )
+      }
```

Comment:
You should be able to do this
```suggestion
      validateRequiredField(priceData, "BaseAsset", isString)
```

Reply:
This doesn't work because it will show `undefined` for class name:
```
 FAIL   xrpl.js  test/models/oracleSet.test.ts
  ● OracleSet › throws w/ missing BaseAsset of PriceDataSeries

    Expected value   "OracleSet: PriceDataSeries must have a `BaseAsset` string"
    Received:
      "undefined: missing field BaseAsset"
```

Reply:
Hm, perhaps the helper function should be updated to support that sort of usecase. If you don't want to then this is fine.

Reply:
I was thinking maybe an additional optional param that contains the `TransactionType`. But yeah that's fair.

Reply:
Yeah I tried this but the error message makes it seem like the `paramName` is at the top level of the `TransactionType` and not sub-level.